In [416]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# to display the total number columns present in the dataset
pd.set_option('display.max_columns', None)
import re

In [417]:
# let's load the titanic dataset
dataset1 = pd.read_csv('bollywood.csv')
dataset2 = pd.read_csv('bollywood - 2.csv')

In [418]:
# let's inspect the first 5 rows
dataset1.head(10)

No      RelDate                   MovieName  ReleaseTime         Genre  \
0   1  Apr-18-2014                    2 States            3  RomanceGenre   
1   2   Jan-4-2013                Table No. 21            4     Thriller    
2   3  Jul-18-2014          Amit Sahni Ki List            4   ComedyGenre   
3   4   Jan-4-2013            Rajdhani Express            4    DramaGenre   
4   5   Jul-4-2014                Bobby Jasoos           44   ComedyGenre   
5   6  May-30-2014                  Citylights            2    DramaGenre   
6   7  Sqp-19-2014               Daawat-E-Ishq            4   ComedyGenre   
7   8  Jan-11-2013  Matru Ki Bijlee Ka Mandola            4   ComedyGenre   
8   9  Jan-10-2014                Dedh Ishqiya            3   ComedyGenre   
9  10  Jan-11-2013                   Gangoobai            4    DramaGenre   

  Budget BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0   36Cr                 104       8576361       26622.0           2527.0   
1   Cr10                  12       1087320        1129.0            137.0   
2   10Cr                   4        572336         586.0             54.0   
3    7Cr                 383         42626          86.0             19.0   
4   18Cr                10.8       3113427        4512.0           1224.0   
5    7Cr                  35       1076591        1806.0             84.0   
6   30Cr                24.6       3905050        8315.0           1373.0   
7   33Cr                  40       2435283        4326.0            647.0   
8   31Cr                  27       2333067        2436.0            591.0   
9    2Cr              Cr0.01          4354           1.0              NaN   

   YoutubeLikeDislikes  
0                29149  
1                 1266  
2                  640  
3                  105  
4                 5736  
5                 1890  
6                 9688  
7                 4973  
8                 3027  
9                    2

In [419]:
# let's inspect the first 5 rows
dataset2.head()

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            N   Romance   
1        52    22-Mar-13                  Aatma            N  Thriller   
2        53     4-Apr-14         Main Tera Hero           LW    Comedy   
3        54    22-Mar-13               Sona Spa            N  Thriller   
4        55    29-Mar-13             Himmatwala            N    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

In [420]:
dataset1.dtypes

No                       int64
RelDate                 object
MovieName               object
ReleaseTime              int64
Genre                   object
Budget                  object
BoxOfficeCollection     object
YoutubeViews             int64
YoutubeLikes           float64
YoutubeDislikes        float64
YoutubeLikeDislikes      int64
dtype: object

In [421]:
dataset2.dtypes

Movie_ID                   int64
Release_Date              object
Movie_Name                object
Release_Time              object
Genre                     object
BudgetCr                   int64
BoxOfficeCollectionCr    float64
Youtube_Views              int64
Youtube_Likes              int64
Youtube_Dislikes           int64
dtype: object

### Data types of similar columns are different for both datasets. If the data has to be merged, type conversion has to be done.

In [422]:
print(dataset1.size)
print(dataset2.size)

561
990


In [423]:
dataset1.head(5)

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos           44   ComedyGenre   18Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361       26622.0           2527.0   
1                  12       1087320        1129.0            137.0   
2                   4        572336         586.0             54.0   
3                 383         42626          86.0             19.0   
4                10.8       3113427        4512.0           1224.0   

   YoutubeLikeDislikes  
0                29149  
1                 1266  
2                  640  
3                  105  
4                 5736

In [424]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            N   Romance   
1        52    22-Mar-13                  Aatma            N  Thriller   
2        53     4-Apr-14         Main Tera Hero           LW    Comedy   
3        54    22-Mar-13               Sona Spa            N  Thriller   
4        55    29-Mar-13             Himmatwala            N    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

In [425]:
# Checking the unique values in the ReleaseTime column.
dataset1.ReleaseTime.unique()

array([ 3,  4, 44,  2,  1])

In [426]:
# Checking the unique values in the Release_Time column.
dataset2.Release_Time.unique()

array(['N', 'LW', 'HS', 'FS'], dtype=object)

### There is an additional category of 44 in dataset1. Checking the number of records with that category

In [427]:
dataset1.groupby(['ReleaseTime']).size()

ReleaseTime
1      3
2     10
3      4
4     33
44     1
dtype: int64

In [428]:
dataset2.groupby(['Release_Time']).size()

Release_Time
FS    14
HS     8
LW    11
N     66
dtype: int64

### There is only 1 record with ReleaseTime as 44. Replacing that with 4, since it could have been an error during data entry

In [429]:
dataset1.loc[dataset1.ReleaseTime == 44, 'ReleaseTime'] = 4

In [430]:
dataset1.groupby(['ReleaseTime']).size()

ReleaseTime
1     3
2    10
3     4
4    34
dtype: int64

### Replace the values of the categorical feature Release_Time from FS, HS, LW and N to 1, 2, 3, 4 in the dataset2 to bring consistency with values in dataset1

In [431]:
dataset2.loc[dataset2.Release_Time == 'FS', 'Release_Time'] = 1
dataset2.loc[dataset2.Release_Time == 'HS', 'Release_Time'] = 2
dataset2.loc[dataset2.Release_Time == 'LW', 'Release_Time'] = 3
dataset2.loc[dataset2.Release_Time == 'N', 'Release_Time'] = 4

In [432]:
dataset2.groupby(['Release_Time']).size()

Release_Time
1    14
2     8
3    11
4    66
dtype: int64

In [433]:
dataset1.shape

(51, 11)

In [434]:
dataset2.shape

(99, 10)

In [435]:
#Printing null rows/columns
null_columns=dataset1.columns[dataset1.isnull().any()]
dataset1[null_columns].isnull().sum()

YoutubeLikes       1
YoutubeDislikes    3
dtype: int64

In [436]:
null_columns=dataset2.columns[dataset2.isnull().any()]
dataset2[null_columns].isnull().sum()

Series([], dtype: float64)

In [437]:
print(dataset1[dataset1.isnull().any(axis=1)][null_columns].head())

Empty DataFrame
Columns: []
Index: [9, 31, 39, 46]


In [438]:
print(dataset1.loc[[9, 31, 39, 46]])

    No      RelDate           MovieName  ReleaseTime        Genre Budget  \
9   10  Jan-11-2013           Gangoobai            4   DramaGenre    2Cr   
31  32  Feb-15-2013            Murder 3            1     Thriller   12Cr   
39  40  Jun-20-2014          Humshakals            2  ComedyGenre   75Cr   
46  47  Mar-15-2013  Mere Dad Ki Maruti            4  ComedyGenre    5Cr   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
9               Cr0.01          4354           1.0              NaN   
31                  25       2249055        1879.0              NaN   
39                63.7       6280868           NaN              1.0   
46                  11        569711        1226.0              NaN   

    YoutubeLikeDislikes  
9                     2  
31                 2626  
39                    2  
46                 1359  


In [439]:
dataset1.YoutubeLikes.fillna(dataset1.YoutubeLikeDislikes - dataset1.YoutubeDislikes, inplace=True )
dataset1.YoutubeDislikes.fillna(dataset1.YoutubeLikeDislikes - dataset1.YoutubeLikes, inplace=True )

In [440]:
print(dataset1[dataset1.isnull().any(axis=1)][null_columns].head())

Empty DataFrame
Columns: []
Index: []


### Converting the data types for YoutubeLikes, YoutubeDislikes and BoxOfficeCollection fields

In [441]:
dataset1['YoutubeLikes'] = dataset1.YoutubeLikes.astype(int)

In [442]:
dataset1['YoutubeDislikes'] = dataset1.YoutubeDislikes.astype(int)

In [443]:
dataset1.head(5)

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos            4   ComedyGenre   18Cr   

  BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
0                 104       8576361         26622             2527   
1                  12       1087320          1129              137   
2                   4        572336           586               54   
3                 383         42626            86               19   
4                10.8       3113427          4512             1224   

   YoutubeLikeDislikes  
0                29149  
1                 1266  
2                  640  
3                  105  
4                 5736

In [444]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            4   Romance   
1        52    22-Mar-13                  Aatma            4  Thriller   
2        53     4-Apr-14         Main Tera Hero            3    Comedy   
3        54    22-Mar-13               Sona Spa            4  Thriller   
4        55    29-Mar-13             Himmatwala            4    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

In [445]:
dataset1['BoxOfficeCollection1'] = pd.to_numeric(dataset1['BoxOfficeCollection'], errors = 'coerce')

In [446]:
print(dataset1[dataset1.isnull().any(axis=1)][null_columns].head())

Empty DataFrame
Columns: []
Index: [9, 15, 49]


In [447]:
print(dataset1.loc[[9, 15, 49]])

    No      RelDate             MovieName  ReleaseTime        Genre Budget  \
9   10  Jan-11-2013             Gangoobai            4   DramaGenre    2Cr   
15  16  Jan-18-2013               Bandook            4  ActionGenre    4Cr   
49  50   May-2-2014  Kya Dilli Kya Lahore            2   DramaGenre    7Cr   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
9               Cr0.01          4354             1                1   
15              Cr0.09         28391             8               14   
49               Cr0.5        492344          2018               64   

    YoutubeLikeDislikes  BoxOfficeCollection1  
9                     2                   NaN  
15                   22                   NaN  
49                 2082                   NaN  


In [448]:
dataset1.loc[9, 'BoxOfficeCollection1'] = 0.01 
dataset1.loc[15, 'BoxOfficeCollection1'] = 0.09 
dataset1.loc[49, 'BoxOfficeCollection1'] = 0.5

In [449]:
print(dataset1.loc[[9, 15, 49]])

    No      RelDate             MovieName  ReleaseTime        Genre Budget  \
9   10  Jan-11-2013             Gangoobai            4   DramaGenre    2Cr   
15  16  Jan-18-2013               Bandook            4  ActionGenre    4Cr   
49  50   May-2-2014  Kya Dilli Kya Lahore            2   DramaGenre    7Cr   

   BoxOfficeCollection  YoutubeViews  YoutubeLikes  YoutubeDislikes  \
9               Cr0.01          4354             1                1   
15              Cr0.09         28391             8               14   
49               Cr0.5        492344          2018               64   

    YoutubeLikeDislikes  BoxOfficeCollection1  
9                     2                  0.01  
15                   22                  0.09  
49                 2082                  0.50  


In [450]:
dataset1 = dataset1.drop(['BoxOfficeCollection'], axis = 1)

In [451]:
dataset1.head(5)

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre   36Cr   
1   2   Jan-4-2013        Table No. 21            4     Thriller    Cr10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre   10Cr   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre    7Cr   
4   5   Jul-4-2014        Bobby Jasoos            4   ComedyGenre   18Cr   

   YoutubeViews  YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  \
0       8576361         26622             2527                29149   
1       1087320          1129              137                 1266   
2        572336           586               54                  640   
3         42626            86               19                  105   
4       3113427          4512             1224                 5736   

   BoxOfficeCollection1  
0                 104.0  
1                  12.0  
2                   4.0  
3                 383.0  
4                  10.8

In [452]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            4   Romance   
1        52    22-Mar-13                  Aatma            4  Thriller   
2        53     4-Apr-14         Main Tera Hero            3    Comedy   
3        54    22-Mar-13               Sona Spa            4  Thriller   
4        55    29-Mar-13             Himmatwala            4    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

In [453]:
for i in range(len(dataset1)):
    dataset1['Budget'].values[i] = re.sub("\D", "", dataset1['Budget'].values[i]) 

In [454]:
dataset1.head()

No      RelDate           MovieName  ReleaseTime         Genre Budget  \
0   1  Apr-18-2014            2 States            3  RomanceGenre     36   
1   2   Jan-4-2013        Table No. 21            4     Thriller      10   
2   3  Jul-18-2014  Amit Sahni Ki List            4   ComedyGenre     10   
3   4   Jan-4-2013    Rajdhani Express            4    DramaGenre      7   
4   5   Jul-4-2014        Bobby Jasoos            4   ComedyGenre     18   

   YoutubeViews  YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  \
0       8576361         26622             2527                29149   
1       1087320          1129              137                 1266   
2        572336           586               54                  640   
3         42626            86               19                  105   
4       3113427          4512             1224                 5736   

   BoxOfficeCollection1  
0                 104.0  
1                  12.0  
2                   4.0  
3                 383.0  
4                  10.8

### Removing the word 'Genre' from all the values in the field 'Genre' in dataset1

In [455]:
dataset1['Genre'] = dataset1['Genre'].str.replace('Genre','')

In [456]:
dataset1.head(5)

No      RelDate           MovieName  ReleaseTime      Genre Budget  \
0   1  Apr-18-2014            2 States            3    Romance     36   
1   2   Jan-4-2013        Table No. 21            4  Thriller      10   
2   3  Jul-18-2014  Amit Sahni Ki List            4     Comedy     10   
3   4   Jan-4-2013    Rajdhani Express            4      Drama      7   
4   5   Jul-4-2014        Bobby Jasoos            4     Comedy     18   

   YoutubeViews  YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  \
0       8576361         26622             2527                29149   
1       1087320          1129              137                 1266   
2        572336           586               54                  640   
3         42626            86               19                  105   
4       3113427          4512             1224                 5736   

   BoxOfficeCollection1  
0                 104.0  
1                  12.0  
2                   4.0  
3                 383.0  
4                  10.8

In [457]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51     4-Jul-14  Lekar Hum Deewana Dil            4   Romance   
1        52    22-Mar-13                  Aatma            4  Thriller   
2        53     4-Apr-14         Main Tera Hero            3    Comedy   
3        54    22-Mar-13               Sona Spa            4  Thriller   
4        55    29-Mar-13             Himmatwala            4    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

### Converting the RelDate field in dataset 1 and dataset 2 from object to datetime format

In [458]:
dataset1['RelDate1'] = pd.to_datetime(dataset1['RelDate'], format='%b-%d-%Y', errors = 'coerce')

In [459]:
dataset1.head(5)

No      RelDate           MovieName  ReleaseTime      Genre Budget  \
0   1  Apr-18-2014            2 States            3    Romance     36   
1   2   Jan-4-2013        Table No. 21            4  Thriller      10   
2   3  Jul-18-2014  Amit Sahni Ki List            4     Comedy     10   
3   4   Jan-4-2013    Rajdhani Express            4      Drama      7   
4   5   Jul-4-2014        Bobby Jasoos            4     Comedy     18   

   YoutubeViews  YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  \
0       8576361         26622             2527                29149   
1       1087320          1129              137                 1266   
2        572336           586               54                  640   
3         42626            86               19                  105   
4       3113427          4512             1224                 5736   

   BoxOfficeCollection1   RelDate1  
0                 104.0 2014-04-18  
1                  12.0 2013-01-04  
2                   4.0 2014-07-18  
3                 383.0 2013-01-04  
4                  10.8 2014-07-04

In [460]:
dataset1 = dataset1.drop(['RelDate'], axis = 1)

In [461]:
dataset2['Release_Date'] = pd.to_datetime(dataset2['Release_Date'], format='%d-%b-%y', errors = 'coerce')

In [462]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51   2014-07-04  Lekar Hum Deewana Dil            4   Romance   
1        52   2013-03-22                  Aatma            4  Thriller   
2        53   2014-04-04         Main Tera Hero            3    Comedy   
3        54   2013-03-22               Sona Spa            4  Thriller   
4        55   2013-03-29             Himmatwala            4    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

In [463]:
dataset1.head()

No           MovieName  ReleaseTime      Genre Budget  YoutubeViews  \
0   1            2 States            3    Romance     36       8576361   
1   2        Table No. 21            4  Thriller      10       1087320   
2   3  Amit Sahni Ki List            4     Comedy     10        572336   
3   4    Rajdhani Express            4      Drama      7         42626   
4   5        Bobby Jasoos            4     Comedy     18       3113427   

   YoutubeLikes  YoutubeDislikes  YoutubeLikeDislikes  BoxOfficeCollection1  \
0         26622             2527                29149                 104.0   
1          1129              137                 1266                  12.0   
2           586               54                  640                   4.0   
3            86               19                  105                 383.0   
4          4512             1224                 5736                  10.8   

    RelDate1  
0 2014-04-18  
1 2013-01-04  
2 2014-07-18  
3 2013-01-04  
4 2014-07-04

In [464]:
dataset1.rename(columns={'No': 'Movie_ID', 
                         'MovieName': 'Movie_Name', 
                         'ReleaseTime': 'Release_Time', 
                         'Genre': 'Genre',
                         'Budget': 'BudgetCr', 
                         'YoutubeViews': 'Youtube_Views', 
                         'YoutubeLikes': 'Youtube_Likes', 
                         'BoxOfficeCollection1': 'BoxOfficeCollectionCr', 
                         'RelDate1': 'Release_Date',
                         'YoutubeDislikes': 'Youtube_Dislikes',
                        }, inplace=True)

In [465]:
dataset1 = dataset1[["Movie_ID", 
                     "Release_Date",
                     "Movie_Name",
                     "Release_Time",
                     "Genre",
                     "BudgetCr",
                     "BoxOfficeCollectionCr",
                     "Youtube_Views",
                     "Youtube_Likes",
                     "Youtube_Dislikes"]]

In [466]:
dataset1.head(5)

Movie_ID Release_Date          Movie_Name  Release_Time      Genre  \
0         1   2014-04-18            2 States             3    Romance   
1         2   2013-01-04        Table No. 21             4  Thriller    
2         3   2014-07-18  Amit Sahni Ki List             4     Comedy   
3         4   2013-01-04    Rajdhani Express             4      Drama   
4         5   2014-07-04        Bobby Jasoos             4     Comedy   

  BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0       36                  104.0        8576361          26622   
1       10                   12.0        1087320           1129   
2       10                    4.0         572336            586   
3        7                  383.0          42626             86   
4       18                   10.8        3113427           4512   

   Youtube_Dislikes  
0              2527  
1               137  
2                54  
3                19  
4              1224

In [467]:
dataset2.head(5)

Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
0        51   2014-07-04  Lekar Hum Deewana Dil            4   Romance   
1        52   2013-03-22                  Aatma            4  Thriller   
2        53   2014-04-04         Main Tera Hero            3    Comedy   
3        54   2013-03-22               Sona Spa            4  Thriller   
4        55   2013-03-29             Himmatwala            4    Action   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        16                   2.50        3998331           2307   
1         8                   8.78         841891            997   
2        40                  55.00        4903941          12657   
3        10                   0.24          79430             61   
4        50                 100.00        2501277              1   

   Youtube_Dislikes  
0               497  
1               306  
2              2139  
3               112  
4                 1

### Now that all the fields and datatypes match for both datasets, merging them both

In [468]:
dataset = pd.concat([dataset1, dataset2], axis=0)

In [469]:
dataset1.shape

(51, 10)

In [470]:
dataset2.shape

(99, 10)

In [471]:
dataset.shape

(150, 10)

In [472]:
dataset.head(10)

Movie_ID Release_Date                  Movie_Name Release_Time      Genre  \
0         1   2014-04-18                    2 States            3    Romance   
1         2   2013-01-04                Table No. 21            4  Thriller    
2         3   2014-07-18          Amit Sahni Ki List            4     Comedy   
3         4   2013-01-04            Rajdhani Express            4      Drama   
4         5   2014-07-04                Bobby Jasoos            4     Comedy   
5         6   2014-05-30                  Citylights            2      Drama   
6         7          NaT               Daawat-E-Ishq            4     Comedy   
7         8   2013-01-11  Matru Ki Bijlee Ka Mandola            4     Comedy   
8         9   2014-01-10                Dedh Ishqiya            3     Comedy   
9        10   2013-01-11                   Gangoobai            4      Drama   

  BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0       36                 104.00        8576361          26622   
1       10                  12.00        1087320           1129   
2       10                   4.00         572336            586   
3        7                 383.00          42626             86   
4       18                  10.80        3113427           4512   
5        7                  35.00        1076591           1806   
6       30                  24.60        3905050           8315   
7       33                  40.00        2435283           4326   
8       31                  27.00        2333067           2436   
9        2                   0.01           4354              1   

   Youtube_Dislikes  
0              2527  
1               137  
2                54  
3                19  
4              1224  
5                84  
6              1373  
7               647  
8               591  
9                 1

In [473]:
print(dataset[dataset.isnull().any(axis=1)][null_columns].head())

Empty DataFrame
Columns: []
Index: [6, 12, 17, 34]


In [474]:
print(dataset.loc[[6, 12, 17, 34]])

    Movie_ID Release_Date             Movie_Name Release_Time     Genre  \
6          7          NaT          Daawat-E-Ishq            4    Comedy   
6         57   2014-08-22               Mardaani            4    Drama    
12        13          NaT             Ek Villain            2   Romance   
12        63   2013-04-19          Ek Thi Daayan            1  Thriller   
17        18          NaT             Filmistaan            2    Comedy   
17        68   2014-05-02           Purani Jeans            2   Romance   
34        35          NaT                Highway            4    Drama    
34        85   2013-06-07  Yamla Pagla Deewana 2            4    Comedy   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
6        30                  24.60        3905050           8315   
6        15                  36.00        4376650          14408   
12       35                 105.50        4518748          13558   
12       24                  45.00        1837437   

In [475]:
dataset.loc[6, 'Release_Date' ]  = pd.to_datetime('Sep-19-2014', format='%b-%d-%Y')
dataset.loc[12, 'Release_Date' ] = pd.to_datetime('Jun-27-2014', format='%b-%d-%Y')
dataset.loc[17, 'Release_Date' ] = pd.to_datetime('Jun-6-2014', format='%b-%d-%Y')
dataset.loc[34, 'Release_Date' ] = pd.to_datetime('Feb-21-2014', format='%b-%d-%Y')

In [476]:
print(dataset[dataset.isnull().any(axis=1)][null_columns].head())

Empty DataFrame
Columns: []
Index: []


In [477]:
dataset.head(20)

Movie_ID Release_Date                  Movie_Name Release_Time      Genre  \
0          1   2014-04-18                    2 States            3    Romance   
1          2   2013-01-04                Table No. 21            4  Thriller    
2          3   2014-07-18          Amit Sahni Ki List            4     Comedy   
3          4   2013-01-04            Rajdhani Express            4      Drama   
4          5   2014-07-04                Bobby Jasoos            4     Comedy   
5          6   2014-05-30                  Citylights            2      Drama   
6          7   2014-09-19               Daawat-E-Ishq            4     Comedy   
7          8   2013-01-11  Matru Ki Bijlee Ka Mandola            4     Comedy   
8          9   2014-01-10                Dedh Ishqiya            3     Comedy   
9         10   2013-01-11                   Gangoobai            4      Drama   
10        11   2014-03-28                 Dishkiyaoon            4     Action   
11        12   2013-01-18                      Inkaar            4    Romance   
12        13   2014-06-27                  Ek Villain            2    Romance   
13        14   2013-01-18               Mumbai Mirror            4    Action    
14        15   2014-08-08               Entertainment            4     Comedy   
15        16   2013-01-18                     Bandook            4     Action   
16        17   2013-01-25                      Race 2            3  Thriller    
17        18   2014-06-06                  Filmistaan            2     Comedy   
18        19   2014-09-12               Finding Fanny            4     Comedy   
19        20   2014-06-13                       Fugly            2     Comedy   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        36                 104.00        8576361          26622   
1        10                  12.00        1087320           1129   
2        10                   4.00         572336            586   
3         7                 383.00          42626             86   
4        18                  10.80        3113427           4512   
5         7                  35.00        1076591           1806   
6        30                  24.60        3905050           8315   
7        33                  40.00        2435283           4326   
8        31                  27.00        2333067           2436   
9         2                   0.01           4354              1   
10       28                   6.00        2397647           3308   
11        5                  10.25        1077723            897   
12       35                 105.50        4518748          13558   
13        4                   1.80         235767            192   
14       60                 163.00        6336058          20472   
15        4                   0.09          28391              8   
16       60                 162.00        7641333          12617   
17       10                   6.00         989096           1608   
18       15                  58.90        6144142          13013   
19       10                  12.60        2343109           2963   

    Youtube_Dislikes  
0               2527  
1                137  
2                 54  
3                 19  
4               1224  
5                 84  
6               1373  
7                647  
8                591  
9                  1  
10               465  
11               114  
12              1094  
13               164  
14              2843  
15                14  
16              2138  
17               173  
18              2746  
19               556

In [478]:
dataset.dtypes

Movie_ID                          int64
Release_Date             datetime64[ns]
Movie_Name                       object
Release_Time                     object
Genre                            object
BudgetCr                         object
BoxOfficeCollectionCr           float64
Youtube_Views                     int64
Youtube_Likes                     int64
Youtube_Dislikes                  int64
dtype: object

In [479]:
dataset['BudgetCr'] = pd.to_numeric(dataset['BudgetCr'])

In [480]:
dataset['Release_Time'] = pd.to_numeric(dataset['Release_Time'])

In [481]:
dataset.head(5)

Movie_ID Release_Date          Movie_Name  Release_Time      Genre  \
0         1   2014-04-18            2 States             3    Romance   
1         2   2013-01-04        Table No. 21             4  Thriller    
2         3   2014-07-18  Amit Sahni Ki List             4     Comedy   
3         4   2013-01-04    Rajdhani Express             4      Drama   
4         5   2014-07-04        Bobby Jasoos             4     Comedy   

   BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
0        36                  104.0        8576361          26622   
1        10                   12.0        1087320           1129   
2        10                    4.0         572336            586   
3         7                  383.0          42626             86   
4        18                   10.8        3113427           4512   

   Youtube_Dislikes  
0              2527  
1               137  
2                54  
3                19  
4              1224

In [482]:
dataset.dtypes

Movie_ID                          int64
Release_Date             datetime64[ns]
Movie_Name                       object
Release_Time                      int64
Genre                            object
BudgetCr                          int64
BoxOfficeCollectionCr           float64
Youtube_Views                     int64
Youtube_Likes                     int64
Youtube_Dislikes                  int64
dtype: object

# Q1. Identify and rectify the eight prominent data quality issues present in the given datasets.    

#### Issue 1 : Duplicate data
#### YoutubeLikeDislikes column in the dataset1 is the sum of the values in two other columns in the same dataset. 
#### It is also not present in the other dataset. Hence removed it from dataset1.

#### Issue 2 :  Invalid data
#### ReleaseTime column in dataset1 has an invalid value (44). Corrected that to 4.

#### Issue 3 : Inconsitent data
#### Release_Time in dataset2 has different values compared to that of ReleaseTime in dataset1. Modified the dataset2
#### as per the mapping [Long Weekend (3), Festive Season (1), Holiday Season (2), and Normal (4)] to make the datasets
#### consistent.

#### Issue 4 : Noise
####  Genre field has the string 'Genre' appended or prepended for all the data values. Similarly, Budget field had 'Cr' added or pretexted for few data values. 

#### Issue 5 : Missing Data
#### YoutubeLikes and YoutubeDislikes fields in dataset1 has 4 null values. Removed those 4 rows from the dataset.

#### Issue 6 : Missing Data
#### The values of Release_Date is missing for 4 records in the dataset1.



# Q2. How many records are present in the dataset? Print the metadata information of dataset. 

In [484]:
dataset.shape

(150, 10)

### There are 150 records in the combined dataset with 10 fields

In [485]:
dataset.dtypes

Movie_ID                          int64
Release_Date             datetime64[ns]
Movie_Name                       object
Release_Time                      int64
Genre                            object
BudgetCr                          int64
BoxOfficeCollectionCr           float64
Youtube_Views                     int64
Youtube_Likes                     int64
Youtube_Dislikes                  int64
dtype: object

In [486]:
dataset.describe()

Movie_ID  Release_Time    BudgetCr  BoxOfficeCollectionCr  \
count  150.000000    150.000000  150.000000             150.000000   
mean    74.840000      3.320000   29.353333              57.863800   
std     43.056242      1.070156   28.164459              97.887547   
min      1.000000      1.000000    2.000000               0.010000   
25%     38.250000      3.000000   11.000000               9.085000   
50%     74.500000      4.000000   21.000000              28.100000   
75%    111.750000      4.000000   35.000000              57.862500   
max    149.000000      4.000000  150.000000             735.000000   

       Youtube_Views  Youtube_Likes  Youtube_Dislikes  
count   1.500000e+02     150.000000        150.000000  
mean    3.342323e+06    7840.400000       1203.080000  
std     3.493044e+06   12713.335141       1847.377283  
min     4.354000e+03       1.000000          1.000000  
25%     1.076874e+06    1379.000000        194.750000  
50%     2.386348e+06    4082.500000        613.500000  
75%     4.542225e+06    9080.250000       1412.250000  
max     2.317107e+07  101275.000000      11888.000000

# Q3. How many movies in each genre got released in different release times?                  

In [487]:
dataset['Genre'].unique

<bound method Series.unique of 0       Romance
1     Thriller 
2        Comedy
3         Drama
4        Comedy
        ...    
94       Comedy
95     Thriller
96       Comedy
97       Comedy
98       Action
Name: Genre, Length: 150, dtype: object>

In [488]:
dataset['Genre'] = dataset['Genre'].str.strip()

In [489]:
dataset.groupby(['Genre','Release_Time']).count()

Movie_ID  Release_Date  Movie_Name  BudgetCr  \
Genre    Release_Time                                                 
Action   1                    3             3           3         3   
         2                    3             3           3         3   
         3                    3             3           3         3   
         4                   15            15          15        15   
Comedy   1                    3             3           3         3   
         2                    5             5           5         5   
         3                    5             5           5         5   
         4                   23            23          23        23   
Drama    1                    4             4           4         4   
         2                    6             6           6         6   
         3                    1             1           1         1   
         4                   24            24          24        24   
Romance  1                    3             3           3         3   
         2                    3             3           3         3   
         3                    4             4           4         4   
         4                   16            16          16        16   
Thriller 1                    4             4           4         4   
         2                    1             1           1         1   
         3                    2             2           2         2   
         4                   22            22          22        22   

                       BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
Genre    Release_Time                                                        
Action   1                                 3              3              3   
         2                                 3              3              3   
         3                                 3              3              3   
         4                                15             15             15   
Comedy   1                                 3              3              3   
         2                                 5              5              5   
         3                                 5              5              5   
         4                                23             23             23   
Drama    1                                 4              4              4   
         2                                 6              6              6   
         3                                 1              1              1   
         4                                24             24             24   
Romance  1                                 3              3              3   
         2                                 3              3              3   
         3                                 4              4              4   
         4                                16             16             16   
Thriller 1                                 4              4              4   
         2                                 1              1              1   
         3                                 2              2              2   
         4                                22             22             22   

                       Youtube_Dislikes  
Genre    Release_Time                    
Action   1                            3  
         2                            3  
         3                            3  
         4                           15  
Comedy   1                            3  
         2                            5  
         3                            5  
         4                           23  
Drama    1                            4  
         2                            6  
         3                            1  
         4                           24  
Romance  1                            3  
         2                            3  
         3                            4  
         4                   

# Q4. Which month of the year, maximum movie releases are seen?                                  

In [490]:
dataset.groupby([dataset['Release_Date'].dt.year, dataset['Release_Date'].dt.month]).agg({'count'})

Movie_ID Release_Date Movie_Name Release_Time Genre  \
                             count        count      count        count count   
Release_Date Release_Date                                                       
2013         1                   8            8          8            8     8   
             2                   6            6          6            6     6   
             3                   8            8          8            8     8   
             4                   4            4          4            4     4   
             5                   6            6          6            6     6   
             6                   4            4          4            4     4   
             7                   9            9          9            9     9   
             8                   4            4          4            4     4   
             9                   5            5          5            5     5   
             10                  6            6          6            6     6   
             11                  5            5          5            5     5   
2014         1                   7            7          7            7     7   
             2                   7            7          7            7     7   
             3                   8            8          8            8     8   
             4                   6            6          6            6     6   
             5                  11           11         11           11    11   
             6                   7            7          7            7     7   
             7                   8            8          8            8     8   
             8                   3            3          3            3     3   
             9                   6            6          6            6     6   
             10                  3            3          3            3     3   
             11                  5            5          5            5     5   
             12                  2            2          2            2     2   
2015         1                   5            5          5            5     5   
             2                   4            4          4            4     4   
             3                   3            3          3            3     3   

                          BudgetCr BoxOfficeCollectionCr Youtube_Views  \
                             count                 count         count   
Release_Date Release_Date                                                
2013         1                   8                     8             8   
             2                   6                     6             6   
             3                   8                     8             8   
             4                   4                     4             4   
             5                   6                     6             6   
             6                   4                     4             4   
             7                   9                     9             9   
             8                   4                     4             4   
             9                   5                     5             5   
             10                  6                     6             6   
             11                  5                     5             5   
2014         1                   7                     7             7   
             2                   7                     7             7   
             3                   8                     8             8   
             4                   6                     6             6   
             5                  11                    11            11   
             6                   7                     7             7   
             7                   8                     8             8   
             8                   3                     3             3   
             9                   6              

# Maximum number of movies for released in the month of May 2014 (11 movies)

# Q5. Which month of the year typically sees most releases of high budgeted movies, that is, movies with budget of 25 crore or more?                                 

In [491]:
# dataset of movies with budget more than 25cr
dataset25cr = dataset[dataset.BudgetCr > 25]
dataset25cr.groupby([dataset25cr['Release_Date'].dt.month]).count()

Movie_ID  Release_Date  Movie_Name  Release_Time  Genre  \
Release_Date                                                            
1                    6             6           6             6      6   
2                    9             9           9             9      9   
3                    6             6           6             6      6   
4                    4             4           4             4      4   
5                    3             3           3             3      3   
6                    5             5           5             5      5   
7                    6             6           6             6      6   
8                    7             7           7             7      7   
9                    4             4           4             4      4   
10                   4             4           4             4      4   
11                   6             6           6             6      6   
12                   2             2           2             2      2   

              BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
Release_Date                                                                  
1                    6                      6              6              6   
2                    9                      9              9              9   
3                    6                      6              6              6   
4                    4                      4              4              4   
5                    3                      3              3              3   
6                    5                      5              5              5   
7                    6                      6              6              6   
8                    7                      7              7              7   
9                    4                      4              4              4   
10                   4                      4              4              4   
11                   6                      6              6              6   
12                   2                      2              2              2   

              Youtube_Dislikes  
Release_Date                    
1                            6  
2                            9  
3                            6  
4                            4  
5                            3  
6                            5  
7                            6  
8                            7  
9                            4  
10                           4  
11                           6  
12                           2

# Februrary month has most number of high budget movies released (9 movies)

# Q6. Which are the top 10 movies with maximum return of investment (ROI)? Calculate ROI as (Box office collection – Budget) / Budget.                                    

In [492]:
dataset['ROI'] = (dataset['BoxOfficeCollectionCr'] - dataset['BudgetCr'])/dataset['BudgetCr']

In [493]:
# Sort the rows of dataframe by column 'ROI'
dataset_sorted_by_roi = dataset.sort_values(by ='ROI', ascending = False)

In [494]:
dataset_sorted_by_roi.head(10)

Movie_ID Release_Date                 Movie_Name  Release_Time    Genre  \
3          4   2013-01-04           Rajdhani Express             4    Drama   
14        65   2013-04-26                 Aashiqui 2             4  Romance   
39        90   2014-12-19                         PK             2    Drama   
82       133   2013-09-13                Grand Masti             3   Comedy   
85       136   2013-09-20               The Lunchbox             4    Drama   
37        88   2013-06-14                     Fukrey             4   Comedy   
8         59   2014-09-05                   Mary Kom             4    Drama   
78       129   2013-10-18                     Shahid             1    Drama   
37        38   2014-07-11  Humpty Sharma Ki Dulhania             4  Romance   
51       102   2013-07-12         Bhaag Milkha Bhaag             4    Drama   

    BudgetCr  BoxOfficeCollectionCr  Youtube_Views  Youtube_Likes  \
3          7                  383.0          42626             86   
14        12                  110.0        2926673           4944   
39        85                  735.0       13270623          43767   
82        35                  298.0        1795640           3015   
85        10                   85.0        1064854           4054   
37         5                   36.2         227912            222   
8         15                  104.0        6086811          26560   
78         6                   40.0        1148516           3155   
37        20                  130.0        6604595          14535   
51        30                  164.0        2635390           4622   

    Youtube_Dislikes        ROI  
3                 19  53.714286  
14               628   8.166667  
39              5693   7.647059  
82               721   7.514286  
85               285   7.500000  
37                39   6.240000  
8               1647   5.933333  
78               396   5.666667  
37              2208   5.500000  
51               456   4.466667

# Above is the list of 10 movies with maximum ROI from the dataset